### ISSUE TITLE
Use `null` instead of `Number.NaN` in NumberField component

### ISSUE THREAD

**Issue by psirenny** [adobe/react-spectrum#6971]
### Provide a general summary of the feature here

Replace use of `Number.NaN` values in the NumberField component with `null`. This would also enable passing in `null` to the NumberField which was requested in: #5524.

### 🤔 Expected Behavior?

The expected behavior is that one must check for `null` before interacting with values returned by a NumberField. This isn't extra work, because one must *already* call `Number.isNaN()` when handling number values.

Why is `Number.NaN` less desirable? Because:
* TypeScript doesn't error if a developer forget to check for NaN
* TypeScript can't determine whether a NaN check was already performed (extra unnecessary checks)
* Serializing a NaN to JSON resolves to a `null` value anyway

### 😯 Current Behavior

All values from a NumberField component need to be checked for `Number.isNaN()`. 

### 🔦 Context

I observe most of my team members override `onChange` and `value` properties to support `null` instead. This leads me to think that the DX of the NumberField component isn't quite right. Developers don't expect a number to not be a number and TypeScript doesn't warn about it.

**Comment by snowystinger** [comment]
Yes, numberfield is a bit weird in this regard. We'll need to see if we can do this without breaking the current code paths. The behavior has been this way for a long time, and if we did away with it, that'd be considered a breaking change, regardless of intent.

**Comment by liam-hunt-bonterra** [comment]
+1 to this. I have a controlled `NumberField` component and I was surprised to find that the empty value is `NaN` rather than `null`. React Aria radios and Selects accept a controlled value of `null`. `null` is more pleasant for type checking as typescript has no `NaN` type.

**Comment by snowystinger** [comment]
> as typescript has no NaN type

This isn't quite true, the type is `number`, which was chosen initially to reduce the need to declare state for the NumberField
```
function ControlledNumberField() {
  const [value, setValue] = useState(5);
  return <NumberField value={value} onChange={setValue}
}
```
vs if we had done null, then you'd need to declare the state as
```
const [value, setValue] = useState<number | null>(5);
```

That said, everything else uses `null` for no value, so I get that people are used to it.

I was thinking about this some more, I'm not sure how we could make this change without it being breaking. We can't return null from onChange because then the types are breaking for the handler.

It would also be breaking on the State context that we pass along for React Aria Context.

It'd be pretty easy to allow it on the value/defaultValue props, if we see `null` come in, we can just reassign to NaN so that changes inside the useNumberFieldState hook are minimal.

We can't depend on a generic type to enable the new behaviour either, since the logical code wouldn't know about that. We also can't assume that the initial value/defaultValue will be null or NaN, it could change to one of those later in which case we'd initially be emitting the wrong value. We could put it behind a prop, though I don't love that, and I don't have a good name for the prop. Open to possible ideas.

Unfortunately though, I think this is going to need to wait for a major.

**Comment by ik-southpole** [comment]
Hi, should we expect further discussion on this, or would you prefer a PR at this point? The fact that NumberField doesn’t accept null prevents us from displaying an empty input, forcing us to prefill it with an odd default value instead of letting the user enter their own.

**Comment by snowystinger** [comment]
If you can think of a way to do this and have it be backwards compatible, then we welcome a PR.

Otherwise, this statement holds
> Unfortunately though, I think this is going to need to wait for a major.

And you can display an empty input by using `NaN`. You can map `null` to `NaN` outside of the component if you are controlling the value.
